exonum-java-binding/core/rust/src/cmd/mod.rs

```diff
@@ -0,0 +1,627 @@
+                public_configs: vec![public_config_path],
+                public_api_address: Some(public_api_address),
+                private_api_address: Some(private_api_address),
+                public_allow_origin: Some(public_allow_origin),
+                private_allow_origin: Some(private_allow_origin),
+            },
+        };
+        finalize.execute()?;
+
+        Ok(node_config_path)
+    }
+}
+
+/// Possible output of the Java Bindings CLI commands.
+pub enum EjbCommandResult {
+    /// Output of the standard Exonum Core commands.
+    Standard(StandardResult),
+    /// Output of EJB-specific `run` command.
+    EjbRun(Config),
+}
+
+impl From<StandardResult> for EjbCommandResult {
+    fn from(result: StandardResult) -> Self {
+        EjbCommandResult::Standard(result)
+    }
+}
+
+/// Interface of Java Bindings CLI commands.
+pub trait EjbCommand {
+    /// Returns the result of command execution.
+    fn execute(self) -> Result<EjbCommandResult, failure::Error>;
+}
+
+/// EJB-specific common (`template`) config format. Includes additional
+/// `supervisor_mode` parameter.
+#[derive(Debug, Serialize, Deserialize)]
+struct EjbTemplateConfig {
+    #[serde(flatten)]
+    standard: CommonConfigTemplate,
+    /// Mode of the Supervisor service.
+    supervisor_mode: SupervisorMode,
+}
+
+impl EjbCommand for GenerateTemplate {
+    fn execute(self) -> Result<EjbCommandResult, failure::Error> {
+        if let StandardResult::GenerateTemplate {
+            template_config_path,
+        } = self.standard.execute()?
+        {
+            // Load standard config file, add supervisor_mode parameter,
+            //   save the EJB-specific common config to the same file.
+            let standard_config: CommonConfigTemplate = load_config_file(&template_config_path)?;
+            let ejb_config = EjbTemplateConfig {
+                standard: standard_config,
+                supervisor_mode: self.supervisor_mode,
+            };
+            save_config_file(&ejb_config, &template_config_path)?;
+
+            Ok(StandardResult::GenerateTemplate {
+                template_config_path,
+            }
+            .into())
+        } else {
+            unreachable!("Standard generate-template command returned invalid result")
+        }
+    }
+}
+
+/// EJB-specific `SharedConfig` (node public config) with an additional
+/// `supervisor_mode` parameter.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct EjbSharedConfig {
+    #[serde(flatten)]
+    standard: SharedConfig,
+    /// Mode of the Supervisor service.
+    supervisor_mode: SupervisorMode,
+}
+
+impl EjbCommand for GenerateConfig {
+    fn execute(self) -> Result<EjbCommandResult, failure::Error> {
+        let ejb_template_config: EjbTemplateConfig =
+            load_config_file(&self.standard.common_config)?;
+        let supervisor_mode = ejb_template_config.supervisor_mode;
+
+        if let StandardResult::GenerateConfig {
```

Comment:
Does anyone consider the reverse-ordered pattern matching readable (first
pattern-match, then perform the operation producing the result)?
I'd consider rewriting to a more readable syntax: execute and then do pattern matching.